ddlite.py

```diff
@@ -1297,6 +1313,355 @@ def show_log(self, idx=None):
+        if __l <= maxlen:
+          _x.append(__x)
+          _y.append(__y)
+          _lengths.append(__l)
+          _w.append(__w)
+      lengths = _lengths
+      x = _x
+      y = _y
+      w = _w
+      if len(lengths) == 0:
+        return None, None, None
+
+    nsamples = len(x)
+    maxlen = np.max(lengths)
+    _x = np.zeros((maxlen, nsamples)).astype('int64')
+    _x_mask = np.zeros((maxlen, nsamples)).astype(theano.config.floatX)
+
+    for idx, sample in enumerate(x):
+      _x[:lengths[idx], idx] = sample
+      _x_mask[:lengths[idx], idx] = 1.
+
+    return _x, _x_mask, y, w
+
+  def transfer_params_from_gpu_to_cpy(self):
+    params = OrderedDict()
+    for k, v in self.lstm_tparams.items():
+      params[k] = v.get_value()
+    return params
+
+  def pred(self, f_pred, data, minibatches):
+    error = 0
+    for id, samples in minibatches:
+      x = [data[1][i] for i in samples]
+      y = [data[2][i] for i in samples]
+      w = [data[3][i] for i in samples]
+      x, mask, y, w = self.process_data(x, y, w, maxlen = None)
+      preds = f_pred(x, mask)
+      error += (preds == y).sum()
+    error = 1. - np.asarray(error, dtype = config.floatX) / len(data[0])
+    return error
+
+  def pred_p(self, f_pred_prob, data, minibatches):
+    error = 0
+    res = []
+    for id, samples in minibatches:
+      l = [data[0][i] for i in samples]
+      x = [data[1][i] for i in samples]
+      y = [data[2][i] for i in samples]
+      w = [data[3][i] for i in samples]
+      x, mask, y, w = self.process_data(x, y, w, maxlen = None)
+      preds = f_pred_prob(x, mask)
+      pred_l = preds.argmax(axis = 1)
+      res.extend([(a, b[1]) for a, b in zip(l, preds)])
+    return res
+
+  def lstm(self,
+           dim = 50,                # word embedding dimension
+           batch_size = 100,        # batch size
+           learning_rate = 0.01,    # learning rate for sgd
+           optimizer = adadelta,    # optimizer
+           epoch = 300,             # learning epoch
+           maxlen = 1000,           # max sequence len
+           dropout = True,
+           verbose = True):
+
+    self.lstm_settings = locals().copy()
+
+    self.lstm_settings['label_dim'] = 2
+    self.lstm_settings['word_size'] = len(self.word_dict)
+    self.init_lstm_params()
+
+    self.init_lstm_theano_params()
+
+    use_noise, x, mask, y, f_pred_prob, f_pred, cost, w = self.build_lstm()
+
+    f_cost = theano.function([x, mask, y, w], cost, name='f_cost')
+    grads = tensor.grad(cost, wrt=list(self.lstm_tparams.values()))
+    f_grad = theano.function([x, mask, y, w], grads, name='f_grad')
+
+    lr = tensor.scalar(name='lr')
+    f_grad_shared, f_update = optimizer(lr, self.lstm_tparams, grads, x, mask, y, w, cost)
+
+    train_data, x, y, w = [], [], [], []
+    lf_probs = self.get_predicted_probability_lf()
+    for idx in self.training():
+        x.append(self.lstm_X[idx])
+        if lf_probs[idx]>0.5:
```

Comment:
No I just mean is there any reason to use e.g. `y=1, w=0.65` versus `y=0, w=0.35` in your code?